versioned_hdf5/subchunk_map.py

```diff
@@ -0,0 +1,299 @@
+
+    >> ds[idx] = arr
+
+    can be translated into
+
+    >> for chunk, arr_idx_raw, index_raw in as_subchunk_map(ds.chunk_size, idx, ds.shape):
+    ..     ds.data_dict[chunk][index_raw] = arr[arr_idx_raw]
+
+    :param chunk_size: the `ChunkSize` of the Dataset
+    :param idx: the "index" to read from / write to the Dataset
+    :param shape: the shape of the Dataset
+    :return: a generator of `(chunk, arr_idx_raw, index_raw)` tuples
+    """
+    assert isinstance(chunk_size, ChunkSize)
+    if isinstance(idx, Tuple):
+        pass
+    elif isinstance(idx, tuple):
+        idx = Tuple(*idx)
+    else:
+        idx = Tuple(ndindex(idx))
+    assert isinstance(shape, tuple)
+
+    if any(dim < 0 for dim in shape):
+        raise ValueError("shape dimensions must be non-negative")
+
+    if len(shape) != len(chunk_size):
+        raise ValueError("chunks dimensions must equal the array dimensions")
+
+    if idx.isempty(shape):
+        # abort early for empty index
+        return
+
+    idx_len: Py_ssize_t = len(idx.args)
+
+    prefix_chunk_size = chunk_size[:idx_len]
+    prefix_shape = shape[:idx_len]
+
+    suffix_chunk_size = chunk_size[idx_len:]
+    suffix_shape = shape[idx_len:]
+
+    chunk_subindexes = []
+
+    n: Py_ssize_t
+    d: Py_ssize_t
+    s: Py_ssize_t
+    i: Slice | IntegerArray | BooleanArray | Integer
+    chunk_idxs: Iterable[Py_ssize_t]
+    chunk_idx: Py_ssize_t
+    chunk_start: Py_ssize_t
+    chunk_stop: Py_ssize_t
+
+    # Process the prefix of the axes which idx selects on
+    for n, i, d in zip(prefix_chunk_size, idx.args, prefix_shape):
+        i = i.reduce((d,))
+
+        # Compute chunk_idxs, e.g., chunk_idxs == [2, 4] for chunk sizes (100, 1000)
+        # would correspond to chunk (slice(200, 300), slice(4000, 5000)).
+        chunk_subindexes_for_axis: list = []
+        if isinstance(i, Slice):
+            if i.step <= 0:
+                raise NotImplementedError(f"Slice step must be positive not {i.step}")
+
+            start: Py_ssize_t = i.start
+            stop: Py_ssize_t = i.stop
+            step: Py_ssize_t = i.step
+
+            if step > n:
+                chunk_idxs = (
+                    (start + k * step) // n
+                    for k in range((stop - start + step - 1) // step)
+                )
+            else:
+                chunk_idxs = range(start // n, (stop + n - 1) // n)
+
+            for chunk_idx in chunk_idxs:
+                chunk_start = chunk_idx * n
+                chunk_stop = min((chunk_idx + 1) * n, d)
+
+                chunk_subindexes_for_axis.append(
+                    (
+                        Slice(chunk_start, chunk_stop, 1),
+                        _subindex_chunk_slice(
+                            chunk_start, chunk_stop, start, stop, step
+                        ),
+                        _subindex_slice_chunk(
+                            start, stop, step, chunk_start, chunk_stop
+                        ),
```

Comment:
This used to be postprocessed by `Slice(...).reduce(d).raw`.
Removing this single line caused runtime to drop from 89ms to 42ms.